bundles/org.openhab.binding.solarforecast/src/main/java/org/openhab/binding/solarforecast/internal/forecastsolar/ForecastSolarObject.java

```diff
@@ -0,0 +1,321 @@
+    private final TreeMap<ZonedDateTime, Double> wattMap = new TreeMap<ZonedDateTime, Double>();
+    private final DateTimeFormatter dateInputFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
+
+    private ZoneId zone = ZoneId.systemDefault();
+    private Optional<String> rawData = Optional.empty();
+    private Instant expirationDateTime;
+
+    public ForecastSolarObject() {
+        expirationDateTime = Instant.now();
+    }
+
+    public ForecastSolarObject(String content, Instant expirationDate) {
+        expirationDateTime = expirationDate;
+        if (!content.isEmpty()) {
+            rawData = Optional.of(content);
+            try {
+                JSONObject contentJson = new JSONObject(content);
+                JSONObject resultJson = contentJson.getJSONObject("result");
+                JSONObject wattHourJson = resultJson.getJSONObject("watt_hours");
+                JSONObject wattJson = resultJson.getJSONObject("watts");
+                String zoneStr = contentJson.getJSONObject("message").getJSONObject("info").getString("timezone");
+                zone = ZoneId.of(zoneStr);
+                Iterator<String> iter = wattHourJson.keys();
+                // put all values of the current day into sorted tree map
+                while (iter.hasNext()) {
+                    String dateStr = iter.next();
+                    // convert date time into machine readable format
+                    try {
+                        ZonedDateTime zdt = LocalDateTime.parse(dateStr, dateInputFormatter).atZone(zone);
+                        wattHourMap.put(zdt, wattHourJson.getDouble(dateStr));
+                        wattMap.put(zdt, wattJson.getDouble(dateStr));
+                    } catch (DateTimeParseException dtpe) {
+                        logger.warn("Exception parsing time {} Reason: {}", dateStr, dtpe.getMessage());
+                        return;
+                    }
+                }
+            } catch (JSONException je) {
+                logger.debug("Error parsing JSON response {} - {}", content, je.getMessage());
+            }
+        }
+    }
+
+    public boolean isValid() {
+        return !(wattHourMap.isEmpty() || wattMap.isEmpty());
+    }
+
+    public boolean isExpired() {
+        return expirationDateTime.isBefore(Instant.now());
+    }
+
+    public double getActualEnergyValue(ZonedDateTime queryDateTime) {
+        if (!isValid()) {
+            return -1;
+        }
+        Entry<ZonedDateTime, Double> f = wattHourMap.floorEntry(queryDateTime);
+        Entry<ZonedDateTime, Double> c = wattHourMap.ceilingEntry(queryDateTime);
+        if (f != null && c == null) {
+            // only floor available
+            if (f.getKey().toLocalDate().equals(queryDateTime.toLocalDate())) {
+                // floor has valid date
+                return f.getValue() / 1000.0;
+            } else {
+                // floor date doesn't fit
+                return -1;
```

Comment:
Consider error handling by throwing and catching an exception.

Reply:
Handling is now based on Excpetions instead of delivering negative values.